packages/icon/package.py

```diff
@@ -1,294 +1,710 @@
+    variant('serialization',
+            default='none',
+            values=('none', ) + serialization_values,
+            description='Enable the Serialbox2 serialization')
 
-    def configure_args(self):
+    # Optimization Features:
+    variant('mixed-precision',
+            default=False,
+            description='Enable mixed precision dycore')
 
-        args = []
+    depends_on('libxml2', when='+coupling')
+    depends_on('libxml2', when='+art')
 
-        # Icon-hammoz:
-        if '+ham' in self.spec:
-            args.append('--enable-atm-phy-echam-submodels')
-            args.append('--enable-hammoz')
+    depends_on('rttov+hdf5', when='+rttov')
+    depends_on('rttov~openmp', when='~openmp+rttov')
 
-        # Serialization
-        if self.spec.variants['serialize_mode'].value != 'none':
-            args.append('--enable-serialization=' +
-                        self.spec.variants['serialize_mode'].value)
+    for x in serialization_values:
+        depends_on('serialbox+fortran~shared',
+                   when='serialization={0}'.format(x))
 
-        # Art
-        if '+art' in self.spec:
-            args.append('--enable-art')
+    depends_on('libcdi-pio+fortran+netcdf', when='+cdi-pio')
+    depends_on('libcdi-pio grib2=eccodes', when='+cdi-pio+grib2')
+    depends_on('libcdi-pio+mpi', when='+cdi-pio+mpi')
 
-        # Claw
-        if '+claw' in self.spec:
-            args.append('--enable-claw')
+    # needed to mimick BB in test.py
+    # only available in icon-exclaim so far
+    depends_on('cdo', when='@exclaim-master,exclaim-test')
 
-        # Eccodes
-        if '+eccodes' in self.spec:
-            args.append('--enable-grib2')
+    default_eccodes = 'eccodes+aec jp2k=openjpeg'
+    depends_on(default_eccodes + ' +fortran', when='+emvorado')
+    depends_on(default_eccodes, when='+grib2~cdi-pio')
+    # Excessive statements to help the concretizer:
+    depends_on(default_eccodes, type='build', when='+grib2+cdi-pio')
 
-        # DACE
-        if '+dace' in self.spec:
-            args.append('--enable-dace')
+    depends_on('yaxt+fortran', when='+cdi-pio')
+    depends_on('lapack')
+    depends_on('blas')
+    depends_on('netcdf-fortran')
 
-        # Ocean
-        if '+ocean' in self.spec:
-            args.append('--enable-ocean')
-        else:
-            args.append('--disable-ocean')
+    depends_on('netcdf-c', when='~cdi-pio')
+    depends_on('netcdf-c', when='+coupling')
+    depends_on('netcdf-c+mpi', when='+parallel-netcdf~cdi-pio')
+    # Excessive statements to help the concretizer:
+    depends_on('netcdf-c+mpi', type='build', when='+parallel-netcdf+cdi-pio')
+
+    default_hdf5 = 'hdf5+szip'
+    depends_on(default_hdf5 + ' +hl+fortran', when='+emvorado')
+    depends_on(default_hdf5, when='+sct')
+    # Excessive statements to help the concretizer:
+    depends_on(default_hdf5, type='build')
+    depends_on(default_hdf5 + ' +mpi', type='build', when='+parallel-netcdf')
```

Comment:
This looks fishy.
The comment says they are excessive statements. But the code makes hdf5 an unconditional build dependency. Requiring hdf5+mpi if +parallel-netcdf doesn't look excessive at all.
I guess
`depends_on(default_hdf5 + ' +mpi', type='build', when='+parallel-netcdf')` should be in the non-excessive section and
`depends_on(default_hdf5, type='build')` is faulty, because MCH runs icon without hdf5 just fine. Maybe we can just remove it.

Reply:
Do you know for why hdf5 is needed at all in ICON?

If not we can get rid of it?

Reply:
No, I don't know.

Reply:
From Sergey: "Note that those statements were needed to work around the old concretizer. depends_on(default_hdf5, type='build') is there not by accident but not needed with the new concretizer and depends_on(default_hdf5 + ' +mpi', type='build', when='+parallel-netcdf') is not needed with the new concretizer either because +parallel-netcdf implies netcdf-c+mpi and netcdf-c+mpi implies hdf5+mpi anyway."